File: coremark-minimal.py
#!/usr/bin/env python3

import os, time
import wasm3

scriptpath = os.path.dirname(os.path.realpath(__file__))
wasm_fn = os.path.join(scriptpath, "./coremark-minimal.wasm")

print("Initializing Wasm3 engine...")

def clock_ms():
    return int(round(time.time() * 1000))

env = wasm3.Environment()
rt = env.new_runtime(2048)

with open(wasm_fn, "rb") as f:
    mod = env.parse_module(f.read())
    rt.load(mod)
    mod.link_function("env", "clock_ms", "i()", clock_ms)

wasm_run = rt.find_function("run")

print("Running CoreMark 1.0...")
res = wasm_run()

if res > 1:
    print(f"Result: {res:.3f}")
else:
    print("Error")

File: coremark-minimal.mjs
#!/usr/bin/env node
//
// Runs coremark-minimal.wasm on the JavaScript engine built into Node. The
// module imports one function, env.clock_ms, and exports run().

import { readFileSync } from "node:fs";

const wasmFile = new URL("./coremark-minimal.wasm", import.meta.url);

const { instance } = await WebAssembly.instantiate(readFileSync(wasmFile), {
    env: {
        clock_ms: () => BigInt(Date.now()),
    },
});

console.log("Running CoreMark 1.01... [should take 12..20 seconds]");

const res = instance.exports.run();

if (res > 1) {
    console.log(`Result: ${res.toFixed(3)}`);
} else {
    console.log("Error");
    process.exit(1);
}
